File: python/problem017.py
#
# Copyright (c) 2015 - buyuk.dev@gmail.com
# ----------------------------------------
# https://projecteuler.net/problem=17
#


num = [None] * 1001
num[0] = "zero"
num[1] = "one"
num[2] = "two"
num[3] = "three"
num[4] = "four"
num[5] = "five"
num[6] = "six"
num[7] = "seven"
num[8] = "eight"
num[9] = "nine"
# ----------------
num[10] = "ten"
num[20] = "twenty"
num[30] = "thirty"
num[40] = "forty"
num[50] = "fifty"
num[60] = "sixty"
num[70] = "seventy"
num[80] = "eighty"
num[90] = "ninety"
# ----------------
num[1000] = "thousand"


for dec in range(10, 100, 10):
    for unit in range(1, 10):
        num[dec + unit] = num[dec] + "-" + num[unit]

num[11] = "eleven"
num[12] = "twelve"
num[13] = "thirteen"
num[14] = "fourteen"
num[15] = "fifteen"
num[16] = "sixteen"
num[17] = "seventeen"
num[18] = "eighteen"
num[19] = "nineteen"

for hun in range(100, 1000, 100):
    for dec in range(0, 100):
        num[hun + dec] = num[hun/100] + " " + "hundred"
        if dec > 0:
            num[hun + dec] += " and " + num[dec]

num[1000] = "one thousand"


def count(ns):
    return len([s for s in ns if s not in " -"])


sum = 0
for i in range(1, 1001):
    sum += count(num[i])

print sum

File: cpp/problem017.cpp
/*
    Copyright (c) 2016 - buyuk.dev@gmail.com
    ----------------------------------------
    Problem 17 - Number letter counts
*/


#include <iostream>
#include <vector>
#include <cctype>
#include <string>


using namespace std;


const int numsnum = 1001;
vector<string> numbers(numsnum);


int count(const string& s)
{
   int counter = 0;
   for (int i = 0; i < s.size(); ++i)
   {
      if (isalpha(s[i]))
      {
         ++counter;
      }
   }
   return counter;
}


int main(int argc, char* argv[]) 
{
   string units[] = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
   for (int i = 0; i < 10; ++i)
   {
      numbers[i] = units[i];
   }

   string tens[] = {"ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"};
   for (int i = 1; i < 10; ++i)
   {
      numbers[i * 10] = tens[i - 1];
   }
   
   for (int ten = 10; ten < 100; ten += 10)
   {
      for (int unit = 1; unit < 10; ++unit)
      {
         numbers[ten + unit] = numbers[ten] + "-" + numbers[unit];
      }
   }
   
   string dozen[] = {"eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"};
   for (int i = 1; i < 10; ++i)
   {
      numbers[10 + i] = dozen[i - 1];
   }
   
   for (int hun = 100; hun < 1000; hun += 100)
   {
      for (int ten = 0; ten < 100; ++ten)
      {
         numbers[hun + ten] = numbers[ten/100] + " " + "hundred";
         if (ten > 0)
         {
            numbers[hun + ten] += " and " + numbers[ten];
         }
      }
   }
   
   numbers[1000] = "one thousand";
   
   int lettersCounter = 0;
   for (int num = 1; num < numsnum; ++num)
   {
      lettersCounter += count(numbers[num]);
   }
   
   cout << "Letters count in numbers up to 1000 excluding zero is " << lettersCounter;
   return 0;
}
